modules/tfm/tfm/boards/common/attest_hal.c

```diff
@@ -0,0 +1,297 @@
+ *
+ * \details The flash in OTP supports only aligned 4 byte write operations.
+ *          This function writes to the encompassing 4 byte word, masking the
+ *          other half-word with 0xFFFF so it is left untouched.
+ *
+ * \param[in]  ptr  Address to write to.
+ * \param[in]  val  Value to write into @p ptr.
+ */
+static void write_halfword(const uint16_t *ptr, uint16_t val)
+{
+	bool top_half = (uint32_t)ptr % 4; /* Addr not div by 4 */
+	uint32_t target_addr = (uint32_t)ptr & ~3; /* Floor address */
+
+	uint32_t val32 = (uint32_t)val | 0xFFFF0000;
+	uint32_t val32_shifted = ((uint32_t)val << 16) | 0x0000FFFF;
+
+	nrfx_nvmc_word_write(target_addr, top_half ? val32_shifted : val32);
+}
+
+
+/* Can later be replaced by a bl_storage call */
+int read_lcs_from_otp(lcs_t *lcs){
+	uint16_t left_assembly = read_halfword(&p_bl_storage_data->lcs.left_assembly);
+	uint16_t left_provisioning =read_halfword(&p_bl_storage_data->lcs.left_provisioning);
+	uint16_t left_secure = read_halfword(&p_bl_storage_data->lcs.left_secure);
+
+	if (left_assembly == STATE_NOT_LEFT) {
+		(*lcs) = ASSEMBLY;
+		return 0;
+	}
+
+	if (left_assembly == STATE_LEFT && left_provisioning == STATE_NOT_LEFT) {
+		(*lcs) = PROVISION;
+		return 0;
+	}
+
+	if (left_provisioning == STATE_LEFT && left_secure == STATE_NOT_LEFT) {
+		(*lcs) = SECURE;
+		return 0;
+	}
+
+	if (left_provisioning == STATE_LEFT) {
+		(*lcs) = DECOMMISSIONED;
+		return 0;
+	}
+
+	/* To reach this the OTP must be corrupted or reading failed */
+	return -EREADLCS;
+}
+
+int update_lcs_in_otp(lcs_t next_lcs){
+	int err;
+	lcs_t current_lcs = 0;
+
+	if(next_lcs == UNKNOWN){
+		return -EINVALIDLCS;
+	}
+
+	err = read_lcs_from_otp(&current_lcs);
+	if (err != 0) {
+		return err;
+	}
+
+	if (next_lcs < current_lcs) {
+		/* Is is only possible to transition into a higher state */
+		return -EINVALIDLCS;
+	}
+
+	if (next_lcs == current_lcs) {
+		/* The same LCS is a valid argument, but nothing to do so return success */
+		return 0;
+	}
+
+	/* As the device starts in ASSEMBLY, it is not possible to write it */
+	if (current_lcs == ASSEMBLY && next_lcs == PROVISION) {
+		write_halfword(&(p_bl_storage_data->lcs.left_assembly), STATE_LEFT);
+		return 0;
+	}
+
+	if (current_lcs == PROVISION && next_lcs == SECURE) {
+		write_halfword(&(p_bl_storage_data->lcs.left_provisioning), STATE_LEFT);
+		return 0;
+	}
+
+	if (current_lcs == SECURE && next_lcs == DECOMMISSIONED) {
+		write_halfword(&(p_bl_storage_data->lcs.left_secure), STATE_LEFT);
+		return 0;
+	}
+
+	/* This will be the case if any invalid transition is tried */
+	return -EINVALIDLCS;
+}
+
+/* End of  temporary solution */
+
+enum tfm_security_lifecycle_t tfm_attest_hal_get_security_lifecycle(void)
+{
+	int err;
+	lcs_t otp_lcs;
+
+
+	err = read_lcs_from_otp(&otp_lcs);
+	if (err != 0) {
+		return TFM_SLC_UNKNOWN;
+	}
+
+	return map_bl_storage_lcs_to_tfm_slc(otp_lcs);
+}
+
+int tfm_attest_update_security_lifecycle_otp(enum tfm_security_lifecycle_t slc){
```

Comment:
Does this function have to be defined as part of the TF-M attestation HAL API?

We will only call this locally. So we can define our own API for this I think.

That way we avoid another noup commit.

Reply:
Good point moved it :)